# PYTHON BASE

### Notebook by

**Emanuele Gravina**<br>
estensione della prima versione offerta da

**Pietro Corte**<br>
Divisione Supporto Informatico (IDE)<br>
Dipartimento Economia e Statistica (ECS)<br>
mailto: pietro.corte@bancaditalia.it

# Introduzione

## Perché Python?

- Facile da imparare, usare, approfondire 
- Multi-paradigma
- Multi-piattaforma
- Ricchissimo di librerie

Utilizzato in molte grandi realtà del mercato informatico come Google, Youtube e RedHat.

La NASA usa Python per lo sviluppo dei sistemi di controllo.



## "Hello World" a confronto

**Java**
```java
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, world!");
    }
}
```

**C**
```c
#include <stdio.h>
main() {
    printf("Hello, world!\n");
}
```

**Python**
```python
print("Hello, world!")
```


> "The joy of coding Python should be in seeing short, concise, readable
classes that express a lot of action in a small amount of clear code<br>
-- not in reams of trivial code that bores the reader to death"
<br><br>*Guido Van Rossum*.

## Un po' di storia

- Creato nel 1991 da Guido Van Rossum
- Prende il nome dai Monty Python
- Coesistono due versioni: 2 e 3
- Le Python Enhancement Proposals ( [PEPS](https://www.python.org/dev/peps) )
- La [PEP 8](https://www.python.org/dev/peps/pep-0008): Style Guide for Python Code
- La documentazione ufficiale: https://www.python.org/doc


### Useremo Python 3!

# Fondamenti del linguaggio

## Python è un linguaggio interpretato

Un programma interprete si occupa di analizzare il codice sorgente presenti in comuni file di testo con estensione  `.py`. Se sintatticamente corretto l'interprete esegue il codice presente nel file.

L'interprete può anche essere utilizzato in modalità interattiva. In questo caso le istruzioni sono eseguite una a una man mano che l'utente le digita al prompt dell'interprete.

In Python, non esiste una fase di compilazione separata, come avviene nei linguaggi compilati (come ad esempio il C), che generi un file eseguibile partendo dal codice sorgente.

## Variabili, espressioni, commenti

In [3]:
''' questa porzione di codice contiene
    variabili, espressioni e commenti '''

x = 101

s = 'io sono una stringa'

# x%2 = resto della divisione per 2 di x
punti = (x/2 + 1)*3 - x%2   # determino il punteggio

sufficiente = punti > 150

print('x/2 =', x/2)
print('x%2 =', x%2)
print('punti =', punti)
print('sufficiente =', sufficiente)

x/2 = 50.5
x%2 = 1
punti = 153.5
sufficiente = True


## Operatori

### Aritmetici
```python
+   -   *   /   **   //   %
```

### Di confronto
```python
==   !=   <   <=   >   >=   in   not in   is   is not    
```

### Booleani
```python
not   and   or
```

### Bit a bit
```python
~   &   |   <<   >>
```

## Operatore ternario

In [5]:
             # espr1   if    espr2    else    espr3   
risultato = 'promosso' if sufficiente else 'bocciato'
print(risultato)

promosso


Il risultato è:
- *espr1* se la condizione *espr2* è vera 
- *espr3* altrimenti

E' fortemente consigliato che *espr1* e *espr3* siano sempre dello stesso tipo!

Da non confondere con l'*istruzione condizionale*

## Tipi di dato

La valutazione di un espressione produce un ***oggetto*** di un certo ***tipo***.
#### I tipi base:

In [6]:
# intero
type(x)

int

In [7]:
# stringa
type(s)

str

In [8]:
# virgola mobile
type(punti)

float

In [9]:
# booleano 
type(sufficiente)

bool

<h3> La lista completa dei tipi built-in</h3>

<table style="margin-left:0 !important">
<thead>
<tr>
<th style="text-align: left">TIPO</th>
<th style="text-align: left">NOME</th>
<th style="text-align: left">DESCRIZIONE</th>
<th style="text-align: left">ESEMPIO</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left">int</td>
<td style="text-align: left">Intero</td>
<td style="text-align: left">Numero intero</td>
<td style="text-align: left">-12, 0, 100, 9999999</td>
</tr>
<tr>
<td style="text-align: left">float</td>
<td style="text-align: left">Reale</td>
<td style="text-align: left">Numero in virgola mobile</td>
<td style="text-align: left">-1.2, 1.1, 5.43e-10, 2.0E110</td>
</tr>
<tr>
<td style="text-align: left">bool</td>
<td style="text-align: left">Booleano</td>
<td style="text-align: left">Valore vero/falso</td>
<td style="text-align: left">True, False</td>
</tr>
<tr>
<td style="text-align: left">complex</td>
<td style="text-align: left">Complesso</td>
<td style="text-align: left">Numero complesso</td>
<td style="text-align: left">1+2j, 1.1+2.3j, 2j</td>
</tr>
<tr>
<td style="text-align: left">str</td>
<td style="text-align: left">Stringa</td>
<td style="text-align: left">Stringa di testo</td>
<td style="text-align: left">'', '''', 'ape', ''l'ape''</td>
</tr>
<tr>
<td style="text-align: left">bytes</td>
<td style="text-align: left">Bytes</td>
<td style="text-align: left">Sequenza di byte</td>
<td style="text-align: left">b'', b'\x00\x01\x02', b'ciao'</td>
</tr>
<tr>
<td style="text-align: left">list</td>
<td style="text-align: left">Lista</td>
<td style="text-align: left">Sequenza mutabile</td>
<td style="text-align: left">[], [1, 1, 2], ['uno', 2, 'tre']</td>
</tr>
<tr>
<td style="text-align: left">tuple</td>
<td style="text-align: left">Tupla</td>
<td style="text-align: left">Sequenza immutabile</td>
<td style="text-align: left">(), (1, 1, 2), ('uno', 2, 'tre')</td>
</tr>
<tr>
<td style="text-align: left">set</td>
<td style="text-align: left">Insieme</td>
<td style="text-align: left">Insieme di oggetti</td>
<td style="text-align: left">{1, 2, 3}, {'uno', 2, 'tre'}</td>
</tr>
<tr>
<td style="text-align: left">dict</td>
<td style="text-align: left">Dizionario</td>
<td style="text-align: left">Mappa chiave valore</td>
<td style="text-align: left">{}, {'uno': 1, 'due': 2, 'tre': 3}</td>
</tr>
</tbody>
</table>

## None

`None` Indica il valore nullo in python. Equivale al `null` del `C` o di `java`.
Il valore `None` si può testare con l'operatore `is`.

In [10]:
X = None
print('X è nullo?:', X is None)
type(X)

X è nullo?: True


NoneType

## Liste

Una lista è una sequenza mutabile di oggetti. 

In [13]:
# inizializzare una lista

lista1 = []
lista2 = list()
lista3 = [3, 'due', 1]

# alcuni dei metodi più usati

lista2.append('due')
lista3.remove('due')
lista3.sort()

print('lista1:', lista1)
print('lista2:', lista2)
print('lista3:', lista3)

lista1: []
lista2: ['due']
lista3: [1, 3]


In [25]:
# altro esempio in cui si odina una lista
alfabeto=["a","b","z","h","c","k","w","d","e"]
alfabeto.sort()
print(alfabeto)

['a', 'b', 'c', 'd', 'e', 'h', 'k', 'w', 'z']


In [26]:
# ordinamento descending con reverse=True
alfabeto.sort(reverse=True)
print(alfabeto)

['z', 'w', 'k', 'h', 'e', 'd', 'c', 'b', 'a']


In [16]:
# Differenza tra extend e append sulle liste
a = [1, 2, 3, 45, 65]
b = [1, 2, 3, 45, 65]

# in questo caso aggiungo un elemento alla lista che è essa stessa una lista
a.append([66, 88])  
print(a)

# stampo ultimo elemento che è una lista in questo caso
print(a[-1])

[1, 2, 3, 45, 65, [66, 88]]
[66, 88]


In [17]:
# in quest'altro caso invece estendo la lista con gli elementi della lista fornita come parametro
b.extend([66, 88])  
print(b)

[1, 2, 3, 45, 65, 66, 88]


In [23]:
# differenza tra del e remove
# del è un'istruzione e ha bisogno dell'inidice dell'elemento
# mentre remove ha bisogno del valore da rimuovere dalla lista

monty = ["spam","eggs","bacon","bob","alice","pluto"]
print("Lista iniziale", monty)

# elimino l'elemento all'indice 0, il primo quindi
del monty[0]

# elimino l'elemento "pluto"
monty.remove("pluto")

print("Lista finale", monty)

Lista iniziale ['spam', 'eggs', 'bacon', 'bob', 'alice', 'pluto']
Lista finale ['eggs', 'bacon', 'bob', 'alice']


In [24]:
# inserisco come primo elemento (indice 0) della lista il valore tolto prima
monty.insert(0,"spam")

# inserisco come sesto elemento (indice 5) della lista il valore tolto prima
# NB: visto che era l'ultimo avrei potuto anche usare .append("pluto")
monty.insert(5,"pluto") 

print("Lista iniziale ripristinata", monty)
print("posizione di pluto",monty.index("pluto"))

Lista iniziale ripristinata ['spam', 'eggs', 'bacon', 'bob', 'alice', 'pluto']
posizione di pluto 5


### Indexing, concatenazione e ripetizione

L'indexing consiste nel poter riferire un elemento tramite un indice. Gli indici delle liste iniziano da 0.

Le liste (come anche le altre sequenze) possone essere concatenate mediante l'operatore "+" e ripetute *n* volte "moltiplicandole" per un intero positivo. 

In [9]:
# accesso ai singoli elementi per indice

lista4 = ['a', 5, 3, 1]

print('secondo elemento di lista4:', lista4[1])
print('ultimo elemento di lista4:', lista4[-1])

secondo elemento di lista4: 5
ultimo elemento di lista4: 1


In [10]:
# concatenazione e ripetizione

frase = 3*['bla'] + ['è', "un'omatopea"]
print(frase)

['bla', 'bla', 'bla', 'è', "un'omatopea"]


### Slicing

Lo slicing è una funzione che restituisce una parte di una sequenza. Ad esempio:

In [11]:
#indici: 0  1      2  3  4
lista = [1, 'uno', 4, 2, 3]

print('dal secondo al terzo    ->', lista[1:3])
print("dall'inizio al quarto   ->", lista[:4])
print('dal quarto alla fine    ->', lista[3:])
print('dal terzo al penultimo  ->', lista[2:-1])
print('gli ultimi due elementi ->', lista[-2:])
print('una copia della lista   ->', lista[:])
print('elementi di indice pari ->', lista[::2])
print('la lista al contrario   ->', lista[::-1])

dal secondo al terzo    -> ['uno', 4]
dall'inizio al quarto   -> [1, 'uno', 4, 2]
dal quarto alla fine    -> [2, 3]
dal terzo al penultimo  -> [4, 2]
gli ultimi due elementi -> [2, 3]
una copia della lista   -> [1, 'uno', 4, 2, 3]
elementi di indice pari -> [1, 4, 3]
la lista al contrario   -> [3, 2, 4, 'uno', 1]


## Stringhe

Una stringa è una sequenza di caratteri immutabile.

In [12]:
s1 = '    io sono una   stringa   '
s2 = "anch'io sono una stringa"
s3 = s2 + ', ma più lunga'
s4 = '''anch'io sono una stringa, ma su
più di una riga'''
s5 = ('pure io sono una stringa su '
      'più di una riga')

# alcuni dei metodi più comuni
s1 = s1.strip()
s2 = s2.upper()
s4 = s4.replace("\n", ' ').replace('ma', 'ero')
s4_tokens = s4.split(', ')
s4_rebuilt = ', '.join(s4_tokens)

print('s1:',s1)
print('s2:',s2)
print('s3:',s3)
print('s4:',s4)
print('s4_tokens:',s4_tokens)
print('s4_rebuilt:', s4_rebuilt)
print('s5:',s5)

s1: io sono una   stringa
s2: ANCH'IO SONO UNA STRINGA
s3: anch'io sono una stringa, ma più lunga
s4: anch'io sono una stringa, ero su più di una riga
s4_tokens: ["anch'io sono una stringa", 'ero su più di una riga']
s4_rebuilt: anch'io sono una stringa, ero su più di una riga
s5: pure io sono una stringa su più di una riga


Anche le stringhe supportano indexing, concatenazione e ripetizione, slicing.

In [13]:
# indexing, concatenazione e ripetizione, slicing

definizione = 'è un sinonimo di dema' + 2*'go'
termine = 'ca' + 3*'po' + 'lo'

print('definizione[0] =', definizione[0])
print('definizione[5:13] =', definizione[5:13])
print(termine, definizione)

definizione[0] = è
definizione[5:13] = sinonimo
capopopolo è un sinonimo di demagogo


### Formattazione

Spesso è utile inserire valori di variabili all'interno di una stringa.<br>
Per far ciò si utilizza il metodo *format*.

In [14]:
nome = 'Chiara'
anni = 25
alta = 1.768

msg = 'Ciao sono {}, ho {} anni e sono alta {:.2f}'.format(nome, anni, alta)
print(msg)

dati = [1.768, 25, 'Chiara']
msg = 'Ciao sono {2}, ho {1} anni e sono alta {0:.2f}'.format(*dati)
print(msg)

# a partire da python 3.6 esistono le f-strings
print(f'Ciao sono {nome}, ho {anni} anni e sono alta {alta:.2f}')

# vecchio stile
msg = 'Ciao sono %s, ho %d anni e sono alta %.2f' % (nome, anni, alta)
print(msg)

Ciao sono Chiara, ho 25 anni e sono alta 1.77
Ciao sono Chiara, ho 25 anni e sono alta 1.77
Ciao sono Chiara, ho 25 anni e sono alta 1.77
Ciao sono Chiara, ho 25 anni e sono alta 1.77


## Tuple

Come le liste anche le tuple sono delle sequenze di oggetti, ma sono immutabili.
Supportano indexing, concatenazione, ripetizione e slicing.

In [15]:
# inizializzare una tupla

tupla1 = () # serve a poco: le tuple sono immutabili"
tupla2 = ('abc', 12, 3.4, [5, 6, 7], 12)
tupla3 = tuple([1, 'due', 3])

# alcuni metodi delle tuple

occorrenze = tupla2.count(12)
posizione1 = tupla2.index(12)
posizione2 = tupla2.index(12, posizione1+1)

print('le occorrenze di 12 sono', occorrenze)
print('la prima occorrenza di 12 è in posizione', posizione1)
print('la seconda occorrenza di 12 è in posizione', posizione2)

le occorrenze di 12 sono 2
la prima occorrenza di 12 è in posizione 1
la seconda occorrenza di 12 è in posizione 4


## Set e frozenset

 I set sono usati per rappresentare un insieme non ordinato di oggetti unici.<br>
 I frozenset sono la versione immutabile dei set.

In [16]:
# inizializzare un set
set1 = set()
set2 = frozenset() # come la tupla vuota non serve a molto
pari = frozenset((2, 4, 6, 8))
primi = {2, 3, 5, 7}

# alcuni metodi
primi.add(7)    # non ha effetto
primi.add(9)    # ops...
primi.remove(9)
primi_pari = primi.intersection(pari)

print('Primi minori di 10:', primi)
print('Unico primo pari: ', primi_pari)
print('"primi_pari" è contenuto in "primi"?', primi_pari.issubset(pari))

Primi minori di 10: {2, 3, 5, 7}
Unico primo pari:  {2}
"primi_pari" è contenuto in "primi"? True


## Dizionari

Un dizionario è costituito da un insieme di coppie chiave-valore.<br>
La chiave deve essere univoca e consente di ottenere instantaneamente il valore ad essa associato.<br>
Solitamente la chiave è una stringa, ma può essere utilizzato un qualsiasi altro tipo a patto che sia immutabile.

In [17]:
# inizializzare un dizionario
d1 = {}
d2 = dict()
kg_conv = {'g': 0.001, 'dag': 0.01, 'hg': 0.1, 'kg': 1, 'mag': 10, 'q': 100}

# esempio di utilizzo
hg2kg = 3*kg_conv['hg']
kg_conv['t'] = 1000

print(f"3 hg equivalgono a {hg2kg:.3f} kg")
print('kg_conv =', kg_conv)
print('um_supportate =', kg_conv.keys())

3 hg equivalgono a 0.300 kg
kg_conv = {'g': 0.001, 'dag': 0.01, 'hg': 0.1, 'kg': 1, 'mag': 10, 'q': 100, 't': 1000}
um_supportate = dict_keys(['g', 'dag', 'hg', 'kg', 'mag', 'q', 't'])


## Funzioni built-in per le sequenze

Di seguito le funzioni più spesso usate nella gestione delle sequenze. Possono essere usate con liste, tuple, stringhe e dizionari. 

In [18]:
# liste

s = [3, 5, 1]
d = {'len':len(s), 'max':max(s), 'min':min(s), 'sorted':sorted(s), 'sum':sum(s)}
print(d)

{'len': 3, 'max': 5, 'min': 1, 'sorted': [1, 3, 5], 'sum': 9}


In [19]:
# stringhe

s = 'Ciao'
d = {'len':len(s), 'max':max(s), 'min':min(s), 'sorted':sorted(s), 'sum':'-'}
print(d)

{'len': 4, 'max': 'o', 'min': 'C', 'sorted': ['C', 'a', 'i', 'o'], 'sum': '-'}


In [20]:
# tuple

s = (2, 9, 7, 1, 5)
d = {'len':len(s), 'max':max(s), 'min':min(s), 'sorted':sorted(s), 'sum':sum(s)}
print(d)

{'len': 5, 'max': 9, 'min': 1, 'sorted': [1, 2, 5, 7, 9], 'sum': 24}


In [21]:
# set

s = {'pesca', 'mela', 'pera'}
d = {'len':len(s), 'max':max(s), 'min':min(s), 'sorted':sorted(s), 'sum':'-'}
print(d)

{'len': 3, 'max': 'pesca', 'min': 'mela', 'sorted': ['mela', 'pera', 'pesca'], 'sum': '-'}


In [22]:
# dizionari

s = {'c': 1, 'a':2, 'f':3}
d = {'len':len(s), 'max':max(s), 'min':min(s), 'sorted':sorted(s), 'sum':'-'}
print('d:', d)

d: {'len': 3, 'max': 'f', 'min': 'a', 'sorted': ['a', 'c', 'f'], 'sum': '-'}


## Istruzione assegnamento

In [23]:
# variabile   =  espressione
mia_variabile = (x + 1)*3 - x**2 + x//25

A differenza di altri linguaggi, in Python l'assegnamento consiste semplicemente nell'associare alla variabile (a sinistra del segno "=") l'*oggetto* risultante dalla valutazione dell'espressione (a destra del segno "=").

L'assegnamento può anche essere multiplo o utilizzato per estrarre elementi dalle tuple o dalle liste; quest'ultima operazione è detta *unpacking*.

In [24]:
# assegnamento multiplo
i = j = 10
a, b = i+1, j+2

# assegnamento per unpacking
nome, anni, _ = ('Chiara', 25, 1.77)

print('a =', a)
print('b =', b)
print('nome =', nome)
print('anni =', anni)

a = 11
b = 12
nome = Chiara
anni = 25


## Istruzione condizionale

In [25]:
if punti < 100:
    esito = 'negativo'
elif punti < 150:
    esito = 'quasi positivo'
elif punti < 200:
    esito = 'positivo'
else:
    lode = True
    esito = 'molto positivo'

print('Dato che il punteggio è', punti, ", l'esito è", esito)

Dato che il punteggio è 153.5 , l'esito è positivo


Per determinare i blocchi di codice in Python si usa l'indentazione, aggiungendo o togliendo 4 spazi per ogni livello. <br> Va bene anche il tab, l'importante è non utilizzarli entrambi!

## Indentazione obbligatoria

L'obbligatorietà dell'indentazione del codice è una caratteristica particolare del linguaggio Python.

In quasi tutti gli altri linguaggi di programmazione i blocchi di codice sono raggruupati usando particolari caratteri (ad esempio le parentesi graffe `{}` di Java) o specifiche istruzioni (come le istruzioni `begin` `end` del Pascal.

In Python i caratteri di spaziatura (spazi e tab) all'inizio di una riga, sono utilizzati per determinare il livello di indentazione di questa riga e di conseguenza i blocchi di codice.

## Cicli iterativi

In Python esistono due modi per iterare l'esecuzione di un certo blocco di codice:

- il ciclo ***for*** : ripete un blocco di codice per ogni elemento di un oggetto iterabile;
- il ciclo ***while*** : ripete un blocco di codice fintantoché una condizione è vera.

### Esempi di for

In [26]:
for i in range(1,4):
    print('Il cubo di', i, 'è:', i**3)

Il cubo di 1 è: 1
Il cubo di 2 è: 8
Il cubo di 3 è: 27


In [18]:
# ciclo da 0 a 4 (5 escluso)
for n in range(5):
    print("ciclo n:", n)

ciclo n: 0
ciclo n: 1
ciclo n: 2
ciclo n: 3
ciclo n: 4


In [19]:
# stampo tutti gi elementi di una lista
primi = [1, 2, 3, 5, 7, 11, 13, 17, 19, 23]
for primo in primi:
    print(primo)

1
2
3
5
7
11
13
17
19
23


In [21]:
#come creare automaticamente una lista numerica grazie a range
lista_numerica = list(range(99, 300))  
print(lista_numerica)

[99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 262, 263, 264, 265, 266, 267, 268, 269, 270, 271, 272, 273, 274, 275, 276, 277, 278, 279, 280, 281, 282, 283, 284, 285, 286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298, 

In [27]:
import random

x = int(input('Inserisci un numero da 1 a 10: '))
tentativi = []
for i in range(10):
    guess = random.randrange(1,11)
    if x == guess:
        print(f'Ho indovinato al {len(tentativi)+1}° tentativo!')
        break    # esco dal ciclo
    if guess in tentativi:
        continue # vado all'iterazione successiva
    tentativi.append(guess)        
else:
    print('Non sono riuscito a indovinare')
        

Inserisci un numero da 1 a 10: 7
Ho indovinato al 4° tentativo!


### Esempi di while

In [28]:
i = 1
while i < 4:
    print('Il cubo di', i, 'è:', i**3)
    i+=1

Il cubo di 1 è: 1
Il cubo di 2 è: 8
Il cubo di 3 è: 27


In [29]:
import random

x = int(input('Inserisci un numero da 1 a 10: '))
guesses = list(range(1, 11))
while len(guesses) > 0:
    j = random.randrange(0, len(guesses))
    guess = guesses.pop(j)
    if x == guess:
        print(f'Ho indovinato al {10-len(guesses)}° tentativo!')
        break # esco dal ciclo
else:
    print('Non sono riuscito a indovinare')

Inserisci un numero da 1 a 10: 6
Ho indovinato al 8° tentativo!


## Funzioni

Le funzioni servono a raggruppare un insieme di istruzioni che eseguono un compito specifico. In questo modo si evita la ripetizione dello stesso codice in più punti del programma rendendolo più ordinato, leggibile, manutenibile e meno incline a errori.

Analogamente alle funzioni matematiche, le funzioni accettano in input zero o più argomenti (o parametri) e restituiscono un risultato.

In [30]:
import math

# definizione di funzione
def area_triangolo(a, b, c): # intestazione
    """ calcola l'area di un triangolo a partire dalle misure dei lati """
    # corpo
    p = (a + b + c) / 2
    q = p * (p - a) * (p - b) * (p - c)
    
    if q <= 0:
        print('I lati in input non sono quelli di un triangolo!')
        return  # resituisce None
    
    area = math.sqrt(q)
    return round(area, 2)

x = area_triangolo(13, 7, 15) # chiamata di funzione
print('area =', x)

area = 45.47


### La documentazione è fondamentale

La stringa che segue l'intestazione della funzione è detta *docstring*. Oltre a commentare ciò che fa la funzione, la *docstring* è il testo che viene mostrato dal comando help dell'interprete python, come anche da ipython digitando il nome della funzione seguito da un punto interrogativo.

Le convenzioni sulle docstring, cosa scrivere o cosa non scrivere con degli esempi sono contenuti nella [PEP 257](https://www.python.org/dev/peps/pep-0257).

I commenti sono una parte molto importante del codice, non solo come docstring, ma in tutti i punti del codice dove possono essere d'aiuto. Soprattutto in linguaggi compatti e altamente espressivi come Python.

Oltre a far comprendere il codice a una persona diversa da quella che lo ha progettato, i commenti sono molto utili anche per chi ha realizzato il codice per poterlo comprendere meglio anche a distanza di tempo. Un codice ben commentato è più manutenibile e riduce la probabilità di introdurre errori nella sua evoluzione durante tutto il suo ciclo di vita.

In [31]:
help(area_triangolo) # oppure con ipython: area_triangolo?

Help on function area_triangolo in module __main__:

area_triangolo(a, b, c)
    calcola l'area di un triangolo a partire dalle misure dei lati



### Passaggio dei parametri

In Python i parametri sono sempre passati *per riferimento*.
Di conseguenza se si modifica un oggetto all'interno di una funzione la modifica permane dopo averla eseguita.

In [32]:
# passaggio per riferimento

def produttoria(lista):
    ''' moltiplica gli elementi di una lista numerica '''
    
    if lista is None or len(lista) == 0:
        return None
    
    t = 1
    while len(lista) > 0:
        t = t * lista.pop()
    return t

l = [1, 2, 3, 4, 5]
p = produttoria(l)

print('l =', l)
print('p =', p)

l = []
p = 120


E' buona regola generale di programmazione evitare quando possibile funzioni di questo tipo e preferire funzioni in cui l'unico valore modificato è quello restituito.

In [33]:
def produttoria2(lista):
    ''' moltiplica gli elementi di una lista numerica '''
    
    if lista is None or len(lista) == 0:
        return None
    
    t = 1
    for e in lista:
        t = t * e
    return t

l = [1, 2, 3, 4, 5]
p = produttoria2(l)

print('l =', l)
print('p =', p)

l = [1, 2, 3, 4, 5]
p = 120


I parametri possono essere passati per posizione o per nome.

In [34]:
def cstr(n, d, s):
    ''' converte un numero "n" in stringa,
        formattandolo con "d" decimali
        e usando il separatore "s" '''
    
    i1 = int(n)
    i2 = n - i1 
    
    if i2 == 0:
        return str(i1)
    
    i2 = int(i2*pow(10,d))
    
    return str(i1) + s + str(i2)
    
print(cstr(3.562131, 2, '.'))
print(cstr(3.562131, s='.', d=2))

3.56
3.56


Il passaggio per nome risulta molto utile quando nella definizione dei parametri ci sono dei valori di default.

In [35]:
def cstr(n, d=2, s='.'):
    ''' converte un numero "n" in stringa,
        formattandolo con "d" decimali
        e usando il separatore "s" '''
    
    i1 = int(n)
    i2 = n - i1 
    
    if i2 == 0:
        return str(i1)
    
    i2 = int(i2*pow(10,d))
    
    return str(i1) + s + str(i2)
    
print(cstr(3.562131))
print(cstr(3.562131, 3))
print(cstr(3.562131, s='|'))

3.56
3.562
3|56


E' anche possibile definire funzioni che abbiano una lista di parametri variabile, sia posizionali che per nome.<br>
Nel caso siano presenti entrambe le tipologie i parametri per nome devono seguire quelli posizionali.

In [36]:
def presentazione(*nomi, **info):
    ''' stampa un messaggio di presentazione '''

    msg = 'Ciao a tutti, sono {}'.format(' '.join(nomi))
    for k, v in info.items():
        if k == 'eta':
            msg += f', ho {v} anni'
        elif k == 'citta':
            msg += f', vengo da {v}'
        elif k == 'colore':
            msg += f', il mio colore preferito è {v}'
    
    print(' e'.join(msg.rsplit(',', 1)))
        
presentazione('Paolo', eta=33)
presentazione('Maria', 'Laura', eta=35, citta='Roma', colore='blu')

Ciao a tutti, sono Paolo e ho 33 anni
Ciao a tutti, sono Maria Laura, ho 35 anni, vengo da Roma e il mio colore preferito è blu


### Scope delle variabili

Tutte le variabili create all’interno di una funzione possono essere usate solo dal codice della funzione stessa. Tali variabili sono dette *locali*.

Le variabili definite all'esterno della funzione sono invece accessibili anche all'interno della funzione e sono dette *globali*.

In [37]:
PI_GRECO = 3.14

def area_ellisse(a1, a2):
    ''' calcola l'area di un elisse dati gli assi '''
    x = a1 / 2
    y = a2 / 2
    print('x =', x)
    return x * y * PI_GRECO

x = 3
a = area_ellisse(5, 10)

print('a =', a)
print('x =', x)
print('y =', y)

x = 2.5
a = 39.25
x = 3


NameError: name 'y' is not defined

### Funzioni ricorsive

Una funzione può richiamare altre funzioni e in particolare anche se stessa. In questo caso si parla di funzione *ricorsiva*.

Affinchè una funzione ricorsiva termini bisogna che esista una *condizione di terminazione* al verificarsi della quale si è in grado di restituire immediatamente il risultato. Il risultato finale viene calcolato riconducendo progressivamente i parametri iniziali ad una situazione in cui la condizione di terminazione risulta vera.

Le ricorsione consente di risolvere dei problemi complessi in poche e semplici linee di codice, ma spesso risulta inefficiente o inapplicabile per l'elevato utilizzo delle risorse del sistema.

In [79]:
# classico esempio del fattoriale

def fatt(x):
    return 1 if x<=1 else x*fatt(x-1)

print('Fattoriale di 10 =', fatt(10))

Fattoriale di 10 = 3628800


In [80]:
print('Fattoriale di 1000 =', fatt(1000))

RecursionError: maximum recursion depth exceeded in comparison

In [78]:
def fatt2(x):
    r = 1
    for i in range(2, x+1):
        r *= i
    return r

print('Fattoriale di 10 =', fatt2(10))
print('Fattoriale di 1000 =', fatt2(1000))

Fattoriale di 10 = 3628800
Fattoriale di 1000 = 4023872600770937735437024339230039857193748642107146325437999104299385123986290205920442084869694048004799886101971960586316668729948085589013238296699445909974245040870737599188236277271887325197795059509952761208749754624970436014182780946464962910563938874378864873371191810458257836478499770124766328898359557354325131853239584630755574091142624174743493475534286465766116677973966688202912073791438537195882498081268678383745597317461360853795345242215865932019280908782973084313928444032812315586110369768013573042161687476096758713483120254785893207671691324484262361314125087802080002616831510273418279777047846358681701643650241536913982812648102130927612448963599287051149649754199093422215668325720808213331861168115536158365469840467089756029009505376164758477284218896796462449451607653534081989013854424879849599533191017233555566021394503997362807501378376153071277619268490343526252000158885351473316117021039681759215109077880193931781

## Comprehension

Le *comprehension* permettono di creare rapidamente nuove liste, set, e dizionari partendo da una sequenza di valori esistenti, eventualmente trasformandone gli elementi e filtrandole.

In [1]:
t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = [x**3 for x in t1]
l2 = [x**3 for x in t1 if x%2 == 0]
s1 = {x**3 for x in t1 if x%2 == 0}
d1 = {f'Cubo di {x}': x**3 for x in t1 if x%2 == 0}

print('t1 =', t1)
print('l1 =', l1)
print('l2 =', l2)
print('s1 =', s1)
print('d1 =', d1)

t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = [27, 64, 1, 64, 125, 1, 512, 216]
l2 = [64, 64, 512, 216]
s1 = {64, 512, 216}
d1 = {'Cubo di 4': 64, 'Cubo di 8': 512, 'Cubo di 6': 216}


## Map e Filter

In Python ci sono due funzioni bult-in che possono svolgere un ruolo simile alle comprehension, che si rifanno a dei costrutti tipici della programmazione funzionale.

In [39]:
def cubo(x):
    return x**3

def pari(x):
    return x%2 == 0

def kv(x):
    return (f'Cubo di {x}', cubo(x))

t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = list(map(cubo, t1))
l2 = list(map(cubo, filter(pari, t1)))
s1 = set(map(cubo, filter(pari, t1)))
d1 = dict(map(kv, filter(pari, t1)))

print('t1 =', t1)
print('l1 =', l1)
print('l2 =', l2)
print('s1 =', s1)
print('d1 =', d1)

t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = [27, 64, 1, 64, 125, 1, 512, 216]
l2 = [64, 64, 512, 216]
s1 = {64, 512, 216}
d1 = {'Cubo di 4': 64, 'Cubo di 8': 512, 'Cubo di 6': 216}


## Funzioni lambda

Le funzioni *lambda* sono delle funzioni anonime, che possono essere definite e utilizzate "al volo".

In [40]:
t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = list(map(lambda x: x**3, t1))
l2 = list(map(lambda x: x**3, filter(lambda x: x%2==0, t1)))
s1 = set(map(lambda x: x**3, filter(lambda x: x%2==0, t1)))
d1 = dict(map(lambda x: (f'Cubo di {x}', x**3), filter(lambda x: x%2==0, t1)))

print('t1 =', t1)
print('l1 =', l1)
print('l2 =', l2)
print('s1 =', s1)
print('d1 =', d1)

t1 = (3, 4, 1, 4, 5, 1, 8, 6)
l1 = [27, 64, 1, 64, 125, 1, 512, 216]
l2 = [64, 64, 512, 216]
s1 = {64, 512, 216}
d1 = {'Cubo di 4': 64, 'Cubo di 8': 512, 'Cubo di 6': 216}


## Le eccezioni

Quando in un programma si verifica un errore viene generata un'*eccezione*.<br>
Le eccezioni si propagano dal blocco di codice dove si verifica verso il blocco di codice o funzione più esterna.<br>
Se un'eccezione si propaga fino al modulo principale senza essere gestita causa l'arresto del programma.

In [41]:
def somma(x, y):
    s = x + y + (x-2)/(x-2) - 1
    return s

def moltiplica(x, y):
    t = 0
    for i in range(0, x):
        t = somma(t, y)
    return t

def dividi(x, y):
    i = 1 / y
    return moltiplica(x, i)
    
x = dividi(3, 2)
print('x =', x)

x = 1.5


In [42]:
dividi(3, 0)

ZeroDivisionError: division by zero

In [43]:
dividi(5, 2)

ZeroDivisionError: float division by zero

### Gestire le eccezioni

Le eccezioni possono essere catturate e gestitite utilzzando il costrutto **try / except**.

In [82]:
# gestione dell'eccezione

def moltiplica(x, y):
    t = 0
    for i in range(0, x):
        try:
            t = somma(t, y)
        except:
            print(f'La funzione somma va in errore per x = {t}')
            t = t + y
        finally:
            print(f'Iterazione n.{i+1}: t = {t}')
    return t

print('dividi(5, 2) =', dividi(5, 2))

Iterazione n.1: t = 0.5
Iterazione n.2: t = 1.0
Iterazione n.3: t = 1.5
Iterazione n.4: t = 2.0
La funzione somma va in errore per x = 2.0
Iterazione n.5: t = 2.5
dividi(5, 2) = 2.5


In [45]:
# gestire diverse eccezioni

def potenza(x, y):
    try:
        r = x**y
    except TypeError:
        print('Parametri non validi')
    except OverflowError:
        print('Impossibile eseguire il calcolo: valore troppo grande')
    except Exception as e:
        print("Errore:", e)
    else:
        print('Il risultato è', r)
        
potenza(2.1, 'a')
potenza(2.1, 999)
potenza(0, -1)
potenza(5, 125)

Parametri non validi
Impossibile eseguire il calcolo: valore troppo grande
Errore: 0.0 cannot be raised to a negative power
Il risultato è 2350988701644575015937473074444491355637331113544175043017503412556834518909454345703125


### Sollevare un'eccezione

Le eccezioni possono essere sollevate intenzionalmente mediante l'istruzione **raise**.

In [46]:
# lanciare intenzionalmente un'eccezione

def potenza(x, y):
    accepted_types = {int, float}
    if type(x) not in accepted_types:
        raise TypeError(f'Parametro non valido: {x}')
    if type(y) not in accepted_types:
        raise TypeError(f'Parametro non valido: {y}')
    
    try:
        r = x**y
    except Exception as e:
        print("Errore:", e)
    else:
        print('Il risultato è', r)

potenza(5, 125)

Il risultato è 2350988701644575015937473074444491355637331113544175043017503412556834518909454345703125


In [47]:
potenza(2.1, 'a')

TypeError: Parametro non valido: a

## Gestione dei file

Per interagire col file system python fornisce la funzione `open` che consente di accedere a un file secondo diverse modalità:

<table align="left">
<tr>
<th style="text-align: left">Modalità</th>
<th style="text-align: left">Descrizione</th>
</tr>
<tr>
<td style="text-align: left"><code>'r'</code></td>
<td style="text-align: left">Apre un file di testo in lettura. Modo di apertura di default dei file.</td>
</tr>
<tr>
<td style="text-align: left"><code>'w'</code></td>
<td style="text-align: left">Apre un file di testo in scrittura.  Se il file non esiste lo crea, altrimenti ne cancella il contenuto.</td>
</tr>
<tr>
<td style="text-align: left"><code>'a'</code></td>
<td style="text-align: left">Apre un file di testo in <em>append</em>.  Il contenuto viene scritto alla fine del file, senza modificare il contenuto esistente.</td>
</tr>
<tr>
<td style="text-align: left"><code>'x'</code></td>
<td style="text-align: left">Apre un file di testo in creazione esclusiva.  Se il file non esiste, restituisce un errore, altrimenti apre in scrittura cancellandone il contenuto.</td>
</tr>
<tr>
<td style="text-align: left"><code>'r+'</code></td>
<td style="text-align: left">Apre un file di testo in modifica. Permette di leggere e scrivere.</td>
</tr>
<tr>
<td style="text-align: left"><code>'w+'</code></td>
<td style="text-align: left">Apre un file di testo in modifica.  Permette di leggere e scrivere e cancella il contenuto del file.</td>
</tr>
</table> 

Queste modalità valgono per i file di testo. Per i file binari per ciascuna modalità bisogna aggiungere la lettera *b* alla fine.

La funzione `open` restituisce un oggetto file che ci consente di eseguire le funzioni per accedere ai dati contenuti e modificarli.

<table align="left">
<tr>
<th style="text-align: left">Metodo</th>
<th style="text-align: left">Descrizione</th>
</tr>
<tr>
<td style="text-align: left"><code>file.read()</code></td>
<td style="text-align: left">Legge e restituisce l’intero contenuto del file come una singola stringa.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.read(n)</code></td>
<td style="text-align: left">Legge e restituisce <code>n</code> caratteri (o byte).</td>
</tr>
<tr>
<td style="text-align: left"><code>file.readline()</code></td>
<td style="text-align: left">Legge e restituisce una riga del file.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.readlines()</code></td>
<td style="text-align: left">Legge e restuisce l’intero contenuto del file come lista di righe (stringhe).</td>
</tr>
<tr>
<td style="text-align: left"><code>file.write(s)</code></td>
<td style="text-align: left">Scrive nel file la stringa <code>s</code> e ritorna il numero di caratteri (o byte) scritti.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.writelines(lines)</code></td>
<td style="text-align: left">Scrive nel file la lista in righe <code>lines</code>.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.tell()</code></td>
<td style="text-align: left">Restituisce la posizione corrente memorizzata dal file object.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.seek(offset)</code></td>
<td style="text-align: left">Modifica la posizione corrente memorizzata dal file object.</td>
</tr>
<tr>
<td style="text-align: left"><code>file.close()</code></td>
<td style="text-align: left">Chiude il file.</td>
</tr>
</table>

Vediamo alcuni esempi con un file avente il seguente contenuto:

```
riga1
riga2
riga3
```

In [57]:
# lettura di una riga

f1 = open('data/file1.txt', 'r')

r1 = f1.readline()
r2 = f1.readline()
f1.close() # il file va sempre chiuso!

print(r2)

riga2



In [59]:
# lettura dell'intero file 1

f1 = open('data/file1.txt', 'r')
for riga in f1:
    print(riga)
f1.close() # il file va sempre chiuso!

riga1

riga2

riga3



In [60]:
# lettura dell'intero file 2

f1 = open('data/file1.txt', 'r')
righe = f1.readlines()
f1.close() # il file va sempre chiuso!

print(righe)

['riga1\n', 'riga2\n', 'riga3\n']


In [61]:
# posizionamento e lettura di alcuni caratteri

f1 = open('data/file1.txt', 'r')
f1.seek(6)
txt = f1.read(5)
f1.close() # il file va sempre chiuso!

print(f'-{txt}-')

-riga2-


In [62]:
# scrittura in coda

f1 = open('data/file1.txt', 'a')
f1.write('riga4\n')
f1.close() # il file va sempre chiuso!

f1 = open('data/file1.txt', 'r')
print(f1.read())
f1.close() # il file va sempre chiuso!

riga1
riga2
riga3
riga4



In [64]:
# ripristiniamo il file iniziale

f1 = open('data/file1.txt', 'w+')
f1.writelines(righe)
f1.seek(0)
print(f1.read())
f1.close() # il file va sempre chiuso!

riga1
riga2
riga3



Che succede se si verificano errori prima di chiudere il file?
Per gestire la cosa in maniera pulita bisognerebbe gestire le eccezioni, ad esempio:

In [65]:
def elab_file(f):
    # ...
    return

f1 = open('data/file1.txt', 'r')
try:
    elab_file(f1)
except Exception as e:
    print("Errore:", e)
finally:
    f1.close()

## Il costrutto *with*

Per risolvere i problemi analoghi a quelli della chiusura del file, python mette a disposizione il costrutto `with`. Con tale costrutto è possibile gestire gli oggetti (come il file) che richiedono l'esecuzione di determinate operazioni all'inizio e/o alla fine del loro utilizzo (`file.close`).

In [66]:
with open('data/file1.txt', 'r') as f1:
    elab_file(f1)

Immediatamente dopo l'apertura del file sarà eseguito il metodo interno `f1.__enter__()` (che nel caso del file non fa niente); mentre dopo l'esecuzione del blocco di codice, anche in presenza di errori, sarà eseguito il metodo interno `f1.__exit__()` (che chiama il metodo `f1.close()`).

## Programmazione a oggetti - cenni

A differenza della classica programmazione procedurale in cui gli elementi organizzativi sono procedure e funzioni, nella programmazione ad oggetti gli elementi organizzativi principali sono gli oggetti. Nell'oggetto confluiscono dati e funzioni che invece nella programmazione procedurale sono separati.


I dati di un oggetto sono memorizzati nelle sue *proprietà* e costituiscono lo *stato * dell'oggetto. L'insieme delle funzioni di un oggetto, dette *metodi*, ne definiscono invece il *comportamento*.

In [67]:
# versione procedurale

def area_cerchio(r):
    return r**2 * math.pi

def circonferenza_cerchio(r):
    return 2 * r * math.pi

raggio = 10
a = area_cerchio(raggio)
c = circonferenza_cerchio(raggio)

print('Area cerchio =', a)
print('Circonferenza cerchio =', c)

Area cerchio = 314.1592653589793
Circonferenza cerchio = 62.83185307179586


In [68]:
# versione a oggetti

class Cerchio: 
    
    def __init__(self, r):
        self.raggio = r
        
    def area(self):
        return self.raggio**2 * math.pi

    def circonferenza(self):
        return 2 * self.raggio * math.pi

mio_cerchio = Cerchio(10) 
a = mio_cerchio.area();
c = mio_cerchio.circonferenza()

print('Raggio cerchio =', mio_cerchio.raggio)
print('Area cerchio =', a)
print('Circonferenza cerchio =', c)

Raggio cerchio = 10
Area cerchio = 314.1592653589793
Circonferenza cerchio = 62.83185307179586


In [1]:
class Veicolo:
    def __init__(self, ruote):
        self.ruote = ruote
        self.velocita = 0
        
    def fermo(self):
        return self.velocita==0
        
    def parti(self, velocita=1):
        if self.fermo():
            self.velocita = velocita
            
    def fermati(self):
        self.velocita = 0
        
    def accelera(self, velocita):
        if velocita <= 0:
            raise ValueError("La velocità dev'essere maggiore di zero")

        if not self.fermo():
            self.velocita += velocita
        
    def rallenta(self, velocita):
        if velocita <= 0:
            raise ValueError("La velocità dev'essere maggiore di zero") 
            
        if not self.fermo() and self.velocita - velocita <= 0:
            self.fermati()
        else:
            self.velocita -= velocita

In [2]:
apecar = Veicolo(3)
print(f'Il veicolo apecar ha {apecar.ruote} ruote')
print(f"L'apecar {'è fermo' if apecar.fermo() else 'si muove'}")

apecar.parti()
print("La velocità dell'apecar è ora", apecar.velocita)

apecar.accelera(20)
print("La velocità dell'apecar è ora", apecar.velocita)

apecar.rallenta(25)
print(f"L'apecar {'è fermo' if apecar.fermo() else 'si muove'}")

Il veicolo apecar ha 3 ruote
L'apecar è fermo
La velocità dell'apecar è ora 1
La velocità dell'apecar è ora 21
L'apecar è fermo


## Duck typing

> "Se sembra un'anatra, nuota come un'anatra e starnazza come un'anatra,<br>
allora probabilmente è un'anatra!"

In Python i tipi sono inferiti dinamicamente, cioè dedotti dal contesto a runtime (*tipizzazione dinamica*). Inoltre i tipi delle variabili possono cambiare durante l'esecuzione del programma (*tipizzazione dinamica debole*).

Per *duck typing* si intende che la semantica di un oggetto è determinata dal suo comportamento e dalle sue proprietà non dal fatto di appartenere a una certa classe.

In [70]:
class Anatra:
    def cammina(self):
        print("Anatra che cammina")
    def nuota(self):
        print("Anatra che nuota")
    def starnazza(self):
        print("Anatra che fa quack!")
        
class Persona:
    def cammina(self):
        print("Persona che imita un'anatra che cammina")
    def nuota(self):
        print("Persona che imita un'anatra che nuota")
    def starnazza(self):
        print("Persona che imita un'anatra che fa quack!")

In [71]:
def vai_anatra(a):
    a.cammina()
    a.nuota()
    a.starnazza()

x = Anatra()
vai_anatra(x)

x = Persona()
vai_anatra(x)

Anatra che cammina
Anatra che nuota
Anatra che fa quack!
Persona che imita un'anatra che cammina
Persona che imita un'anatra che nuota
Persona che imita un'anatra che fa quack!


## I moduli

Il modulo è un insieme costanti, funzioni e classi raggruppati in un file. I moduli consentono di suddividere e organizzare meglio il codice.

Python include già un'ampia lista di moduli standard (python standard library), ma è anche possibile scaricarne o definirne di nuovi. Ad esempio uno dei moduli standard è il modulo `math` che contiene diverse costanti e funzioni matematiche.

Tutti i moduli disponibili nella libreria standard sono descritti nella [documentazione ufficiale](https://docs.python.org/3.7/library/index.html).

Per poter uitilizzare gli elementi contenuti in un certo modulo è necessario *importarlo*.

In [72]:
import math

print('Pi greco vale', math.pi)
print('Il seno di 90 gradi è', math.sin(math.pi/2))

Pi greco vale 3.141592653589793
Il seno di 90 gradi è 1.0


Si possono anche importare solo i nomi che interessano e usarli direttamente. 

In [73]:
from math import sqrt

print('La radice quadrata di 16 è', sqrt(16))

La radice quadrata di 16 è 4.0


Si può rinominare ciò che si importa per evitare conflitti di nomi o semplicemente per comodità.

In [74]:
import datetime as dt

print('Oggi è il', dt.date.today().strftime('%d-%m-%Y'))

Oggi è il 11-02-2021


Per ottenere tutti i nomi definiti in un modulo si può usare il comando `dir`.

In [75]:
dir(dt)

['MAXYEAR',
 'MINYEAR',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'date',
 'datetime',
 'datetime_CAPI',
 'time',
 'timedelta',
 'timezone',
 'tzinfo']

Se i nomi non sono molti e non vanno in conflitto, si possono importare tutti i nomi *visibili* (cioè che non iniziano con `_` ) di un modulo. Ad esempio:

In [76]:
from datetime import *

domani = date.today() + timedelta(days=1)
print('Domani è il', domani.strftime('%d-%m-%Y'))

Domani è il 12-02-2021


`dir()` senza parametri mostra i nomi visibili dallo script/modulo corrente.

### Creare un nuovo modulo

Per creare un nuovo modulo è sufficiente definire gli elementi che lo costituiscono (costanti, funzioni e oggetti) all'interno di uno stesso file di testo e salvarlo col nome *&lt;nome_modulo&gt;.py*.

Affinchè un modulo (non facente parte di un *package*) sia individuabile dall'interprete python e *importabile* il percorso della cartella che lo contiene deve trovarsi nella variabile d'ambiente *PYTHONPATH* opure il modulo deve trovarsi nella directory corrente.

Non c'è nessuna distinzione, se non quella logica, tra un modulo e un file contenente del codice python che si vuole eseguire.

## I package

Un *package* python è una raccolta di moduli organizzati in maniera gerarchica a partire (quasi sempre) da una stessa cartella del filesystem. La cartella e le eventuali sottocartelle appartenenti al *package* devono contenere oltre ai moduli il file `__init__.py` (anche vuoto).

Tale file è destinato a contenere l'eventuale codice di inizializzazione del package o la semplice definizione della variabile speciale `__ALL__`.

Questa variabile è una sequenza contenente la lista dei nomi definiti all'interno dei moduli del package ed visibili all'esterno.

Se la variabile `__ALL__` non è definita tutti i nomi che non iniziano per `'_'` risultano visibili.

Affinchè un package (con tutti i suoi moduli) sia individuabile dall'interprete python e *importabile* il percorso della cartella che lo contiene deve trovarsi nella variabile d'ambiente *PYTHONPATH* oppure il package deve trovarsi nella directory corrente.

### Esempio di struttura di un package

<br>

```
miopack/                     Package principale
      __init__.py            Inizializza il package miopack
      mod1.py
      sub1/                  Sottopackage sub1
            __init__.py
            mod11.py
            mod12.py
      sub2/                  Sottopackage sub2
            __init__.py
            mod21.py
            mod22.py
            mod23.py
```

## Il Python Package Index e pip

Il Python Package Index (`PyPI`) è un repository che contiene decine di migliaia di package scritti in Python. Chiunque può scaricare package esistenti o condividere nuovi package su PyPI. PyPI è anche conosciuto con il nome di Cheese Shop, dallo sketch del cheese shop di Monty Python.

È possibile accedere ai package del Python Package Index sia tramite un browser (all’indirizzo https://pypi.org), sia tramite un il comando `pip`.

`pip` è un comando che ci consente di cercare, scaricare ed installare package Python che si trovano sul Python Package Index. Il nome è un acronimo ricorsivo, che significa Pip Installs Packages. `pip` ci consente inoltre di gestire i package che abbiamo già scaricato, permettendonci di aggiornarli o rimuoverli.

In [77]:
!pip help


Usage:   
  pip <command> [options]

Commands:
  install                     Install packages.
  download                    Download packages.
  uninstall                   Uninstall packages.
  freeze                      Output installed packages in requirements format.
  list                        List installed packages.
  show                        Show information about installed packages.
  check                       Verify installed packages have compatible dependencies.
  config                      Manage local and global configuration.
  search                      Search PyPI for packages.
  wheel                       Build wheels from your requirements.
  hash                        Compute hashes of package archives.
  completion                  A helper command used for command completion.
  debug                       Show information useful for debugging.
  help                        Show help for commands.

General Options:
  -h, --help                  Show h